File: javascript/easy/intersectionOfTwoLinkedLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    let currentA = headA, currentB = headB, sizeA = getSize(headA), sizeB = getSize(headB);
    if(sizeA > sizeB) {
        currentA = moveNode(currentA, (sizeA - sizeB));
    } else {
        currentB = moveNode(currentB, (sizeB - sizeA));
    }

    while(currentA != null && currentB != null) {
        if(currentA == currentB) {
            return currentA;
        }
        currentA = currentA.next;
        currentB = currentB.next;
    }

    return null;
};

var getSize = function(head) {
    let size = 0;
    while(head != null) {
        head = head.next;
        size++;
    }
    return size;
};

var moveNode = function(head, number) {
    let i = 0;
    while(head != null && i < number) {
        head = head.next;
        i++;
    }
    return head;
};

File: ruby/easy/intersection_of_two_linked_lists.rb
# # Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

def get_intersection_node(a, b)
    len_a = len(a)
    len_b = len(b)
    if len_a > len_b
        while(len_b < len_a)
            a = a.next
            len_b += 1
        end
    elsif len_b > len_a
        while(len_a < len_b)
            b = b.next
            len_a += 1
        end
    end
    while a && b
        return a if a == b
        a = a.next
        b = b.next
    end
    return nil
end

def len(n)
    len = 0
    while(n)
        len += 1
        n = n.next
    end
    return len
end
